hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/KeyManagerImpl.java

```diff
@@ -942,17 +944,159 @@ private boolean isKeyEmpty(OmKeyInfo keyInfo) {
     // snapshot of the data, so we don't need these locks at a higher level
     // when we iterate.
 
+    boolean preserveTrailingSlash = enableFileSystemPaths &&
+                    !OzoneManagerRatisUtils.isOmLayoutVersionV1();
     if (enableFileSystemPaths) {
-      startKey = OmUtils.normalizeKey(startKey, true);
-      keyPrefix = OmUtils.normalizeKey(keyPrefix, true);
+      startKey = OmUtils.normalizeKey(startKey, preserveTrailingSlash);
+      keyPrefix = OmUtils.normalizeKey(keyPrefix, preserveTrailingSlash);
+    }
+
+    validateListKeyArgs(volumeName, bucketName);
+    List<OmKeyInfo> keyList = new ArrayList<>();
+    if (maxKeys <= 0) {
+      return keyList;
+    }
+
+    if (OzoneManagerRatisUtils.isOmLayoutVersionV1()) {
+      return listKeysV1(volumeName, bucketName, startKey, keyPrefix, maxKeys);
+    }
+
+    keyList = metadataManager.listKeys(volumeName, bucketName, startKey,
+            keyPrefix, maxKeys);
+
+    return keyList;
+  }
+
+  private List<OmKeyInfo> listKeysV1(String volumeName, String bucketName,
+      String startKey, String keyPrefix, int maxKeys) throws IOException {
+    List<OmKeyInfo> keyList = new ArrayList<>();
+    if (StringUtils.isBlank(startKey)) {
+      return listKeysRecursivelyFromKeyPrefixV1(volumeName, bucketName,
+              keyPrefix, maxKeys, keyList);
+    } else {
+      return listKeysRecursivelyFromStartKeyV1(volumeName, bucketName, startKey,
+              keyPrefix, maxKeys, keyList);
+    }
+  }
+
+  private List<OmKeyInfo> listKeysRecursivelyFromKeyPrefixV1(String volumeName,
+      String bucketName, String keyPrefix, int maxKeys,
+      List<OmKeyInfo> keyList) throws IOException {
+
+    String startKey = "";
+    int countEntries = keyList.size();
+    List<OzoneFileStatus> keys = listOzoneFileStatusesV1(false,
+            startKey, countEntries, maxKeys, null, volumeName,
+            bucketName, keyPrefix, true, true);
+
+    for (OzoneFileStatus status : keys) {
+      OmKeyInfo omKeyInfo = status.getKeyInfo();
+      keyList.add(omKeyInfo);
+      if (status.isDirectory()) {
+        // add trailing slash to represent directory
+        String dirName = omKeyInfo.getKeyName();
+        if (!dirName.endsWith("/")) {
+          omKeyInfo.setKeyName(dirName + OM_KEY_PREFIX);
+        }
+        if (!StringUtils.equals(dirName, keyPrefix)) {
+          listKeysRecursivelyFromKeyPrefixV1(volumeName, bucketName, dirName,
+                  maxKeys, keyList);
```

Comment:
`maxKeys` represents the limit and the batch size. During iteration and seeks, keyList.size() shouldn't exceed this limit. So, maxKeys is static.

`keyList.size()  <= maxKeys`

Reply:
Okay, one improvement I am thinking: can we do a check if keyList.size() already reach maxKeys, if reached, no need to call listKeysRecursivelyFromKeyPrefixV1 for dir. This can avoid unnecessary listKeysRecursivelyFromKeyPrefixV1 be invoked for subsequent sub-dirs.

Reply:
Agreed, will do the check.